File: samples/src/main/java/com/compdfkit/samples/samples/FlattenTest.java
/**
 * Copyright © 2014-2025 PDF Technologies, Inc. All Rights Reserved.
 *
 * THIS SOURCE CODE AND ANY ACCOMPANYING DOCUMENTATION ARE PROTECTED BY INTERNATIONAL COPYRIGHT LAW
 * AND MAY NOT BE RESOLD OR REDISTRIBUTED. USAGE IS BOUND TO THE ComPDFKit LICENSE AGREEMENT.
 * UNAUTHORIZED REPRODUCTION OR DISTRIBUTION IS SUBJECT TO CIVIL AND CRIMINAL PENALTIES.
 * This notice may not be removed from this file.
 */

package com.compdfkit.samples.samples;

import com.compdfkit.core.document.CPDFDocument;
import com.compdfkit.core.page.CPDFPage;
import com.compdfkit.samples.PDFSamples;
import com.compdfkit.samples.R;
import com.compdfkit.samples.util.FileUtils;
import com.compdfkit.samples.util.OutputListener;

import java.io.File;


public class FlattenTest extends PDFSamples {

    public FlattenTest(){
        setTitle(R.string.flatten_test_title);
        setDescription(R.string.flatten_test_desc);
    }

    @Override
    protected void run(OutputListener outputListener) {
        super.run(outputListener);
        printHead();
        outputListener.println();
        CPDFDocument document = new CPDFDocument(context);
        document.open(FileUtils.getAssetsTempFile(context, "Annotations.pdf"));
        // Get the total number of comments in the pdf document
        int annotationCount = 0;
        for (int i = 0; i < document.getPageCount(); i++) {
            annotationCount += document.pageAtIndex(i).getAnnotCount();
        }
        outputListener.println(annotationCount + " annotations in the file.");
        // Flatten processing of pdf pages
        document.flattenAllPages(CPDFPage.PDFFlattenOption.FLAT_NORMALDISPLAY);
        // save document
        File file = new File(outputDir(),
                "Flatten/FlattenTest.pdf");
        saveSamplePDF(document, file, true);
        outputListener.println("Done. Result saved in FlattenTest.pdf");

        CPDFDocument newFileDocument = new CPDFDocument(context);
        newFileDocument.open(file.getAbsolutePath());
        //Open the document again and get the number of annotations in the document
        annotationCount = 0;
        for (int i = 0; i < newFileDocument.getPageCount(); i++) {
            annotationCount += newFileDocument.pageAtIndex(i).getAnnotCount();
        }
        newFileDocument.close();
        outputListener.println(annotationCount + " annotations in the new file.");

        printDividingLine();
        printFooter();
    }
}

File: Samples_kotlin/src/main/java/com/compdfkit/samples/samples/FlattenTest.kt
/**
 * Copyright © 2014-2025 PDF Technologies, Inc. All Rights Reserved.
 *
 * THIS SOURCE CODE AND ANY ACCOMPANYING DOCUMENTATION ARE PROTECTED BY INTERNATIONAL COPYRIGHT LAW
 * AND MAY NOT BE RESOLD OR REDISTRIBUTED. USAGE IS BOUND TO THE ComPDFKit LICENSE AGREEMENT.
 * UNAUTHORIZED REPRODUCTION OR DISTRIBUTION IS SUBJECT TO CIVIL AND CRIMINAL PENALTIES.
 * This notice may not be removed from this file.
 */
package com.compdfkit.samples.samples

import com.compdfkit.core.document.CPDFDocument
import com.compdfkit.core.page.CPDFPage
import com.compdfkit.samples.PDFSamples
import com.compdfkit.samples.R
import com.compdfkit.samples.util.FileUtils.getAssetsTempFile
import com.compdfkit.samples.util.OutputListener
import java.io.File

class FlattenTest : PDFSamples() {
    init {
        setTitle(R.string.flatten_test_title)
        setDescription(R.string.flatten_test_desc)
    }

    override fun run(outputListener: OutputListener?) {
        super.run(outputListener)
        printHead()
        outputListener?.println()
        val document = CPDFDocument(context())
        document.open(getAssetsTempFile(context(), "Annotations.pdf"))
        // Get the total number of comments in the pdf document
        var annotationCount = 0
        for (i in 0 until document.pageCount) {
            annotationCount += document.pageAtIndex(i).annotCount
        }
        outputListener?.println("$annotationCount annotations in the file.")
        // Flatten processing of pdf pages
        document.flattenAllPages(CPDFPage.PDFFlattenOption.FLAT_NORMALDISPLAY)
        // save document
        val file = File(outputDir(), "Flatten/FlattenTest.pdf")
        saveSamplePDF(document, file, true)
        outputListener?.println("Done. Result saved in FlattenTest.pdf")
        val newFileDocument = CPDFDocument(context())
        newFileDocument.open(file.absolutePath)
        //Open the document again and get the number of annotations in the document
        annotationCount = 0
        for (i in 0 until newFileDocument.pageCount) {
            annotationCount += newFileDocument.pageAtIndex(i).annotCount
        }
        newFileDocument.close()
        outputListener?.println("$annotationCount annotations in the new file.")
        printDividingLine()
        printFooter()
    }
}